### PR TITLE
Add blog post "Stateful Processing with Apache Beam"

### PR DESCRIPTION
**Pull request by kennknowles**
This blog post introduces stateful processing in Beam, after a bit of a recap of the basics of the model.

### CODE REVIEW

File: src/_posts/2017-02-09-stateful-processing.md [hunk 1]
```diff
@@ -0,0 +1,524 @@
...
+specified by `"index"` _along with_ the key and window currently being
+processed.  So, when thinking about a state cell, it may be helpful to consider
+the full state of your transform as a table, where the rows are named according
+to names you use in your program, like `"index"`, and the columns are
+key+window pairs, like this:
+
+|               | (key, window)<sub>1</sub> | (key, window)<sub>2</sub> | (key, window)<sub>3</sub> | ... |
+|---------------|---------------------------|---------------------------|---------------------------|-----|
+| `"index"`       | `3`                         | `7`                         | `15`                        | ... |
+| `"fizzOrBuzz?"` | `"fizz"`                    | `"7"`                       | `"fizzbuzz"`                | ... |
```
Comment (aljoscha) [inline src/_posts/2017-02-09-stateful-processing.md]: 😂

**Comment by davorbonaci** [comment]
```
- ./content/subdir/documentation/pipelines/create-your-pipeline/index.html
  *  linking to /subdir/documentation/programming-guide/#transforms-composite, but transforms-composite does not exist (line 281)
     <a href="/subdir/documentation/programming-guide/#transforms-composite">composite transforms</a>
- ./content/subdir/documentation/pipelines/test-your-pipeline/index.html
  *  linking to /subdir/documentation/programming-guide/#transforms-composite, but transforms-composite does not exist (line 186)
     <a href="/subdir/documentation/programming-guide/#transforms-composite">Composite Transform</a>
- ./content/subdir/get-started/mobile-gaming-example/index.html
  *  linking to /subdir/documentation/programming-guide/#transforms-composite, but transforms-composite does not exist (line 255)
     <a href="/subdir/documentation/programming-guide/#transforms-composite">user-defined composite transform</a>
```

**Comment by kennknowles** [comment]
That was breakage in master when I pushed. I think the fix is in.

**Comment by kennknowles** [comment]
Rebased so now build is fine. Makes sense to hold for Monday.

File: src/_posts/2017-02-09-stateful-processing.md [hunk 2]
```diff
@@ -0,0 +1,524 @@
...
+efficiencies. In this post, I will guide you through stateful processing in
+Beam: how it works, how it fits in with the other features of the Beam model,
+what you might use it for, and what it looks like in code.
+
+<!--more-->
+
+> **Warning: new features ahead!**: This is a very new aspect of the Beam
+> model. Runners are still adding support. You can try it out today on multiple
+> runners, but do check the [runner capability
+> matrix](https://beam.apache.org/documentation/runners/capability-matrix/) for
```
Comment (davorbonaci) [inline src/_posts/2017-02-09-stateful-processing.md]: relative link

File: src/_posts/2017-02-09-stateful-processing.md [hunk 3]
```diff
@@ -0,0 +1,524 @@
...
+   lateness) but this could mean that the runner is tracking an additional timer
+   per key and window to cause reclamation code to execute.
+
+## Go use it!
+
+If you are new to Beam, I hope you are now interested in seeing if Beam with
+stateful processing addresses your use case.  If you are already using Beam, I
+hope this new addition to the model unlocks new use cases for you.  Do check
+the [capability
+matrix](https://beam.apache.org/documentation/runners/capability-matrix/) to
```
Comment (davorbonaci) [inline src/_posts/2017-02-09-stateful-processing.md]: relative link

File: src/_posts/2017-02-09-stateful-processing.md [hunk 4]
```diff
@@ -0,0 +1,524 @@
...
+see the level of support for this new model feature on your favorite
+backend(s). 
+
+And please do join the community at
+[user@beam.apache.org](mailto:user-subscribe@beam.apache.org). We'd love to
```
Comment (davorbonaci) [inline src/_posts/2017-02-09-stateful-processing.md]: Link to https://beam.apache.org/get-started/support/

File: src/_posts/2017-02-09-stateful-processing.md [hunk 5]
```diff
@@ -0,0 +1,524 @@
...
+
+## Example: arbitrary-but-consistent index assignment
+
+Suppose that you want to give an index to every incoming element for a
+key-and-window. You don't care what the indices are, just as long as they are
+unique and consistent. Before diving into the code for how to do this in a Beam
+SDK, I'll go over this example from the level of the model. In pictures, you
+want to write a transform that maps input to output like this:
+
+<img class="center-block" src="{{
```
Comment (davorbonaci) [inline src/_posts/2017-02-09-stateful-processing.md]: change style

File: src/_posts/2017-02-09-stateful-processing.md [hunk 6]
```diff
@@ -0,0 +1,524 @@
...
+
+## State in Beam's Java SDK
+
+Now that I have talked a bit about stateful processing in the Beam model and
+worked through an abstract example, I'd like to show you what it looks like to
+write stateful processing code using Beam's Java SDK.  Here is the code for a
+stateful `DoFn` that assigns an arbitrary-but-consistent index to each element
+on a per key-and-window basis:
+
+```java
```
Comment (davorbonaci) [inline src/_posts/2017-02-09-stateful-processing.md]: may not show up if somewhere else python is clicked

File: src/_posts/2017-02-09-stateful-processing.md (hunk 2 above)
Comment (kennknowles) [inline src/_posts/2017-02-09-stateful-processing.md]: done

File: src/_posts/2017-02-09-stateful-processing.md [hunk 7]
```diff
@@ -0,0 +1,524 @@
+---
+layout: post
+title:  "Stateful processing with Apache Beam"
+date:   2017-02-09 00:00:01 -0800
```
Comment (kennknowles) [inline src/_posts/2017-02-09-stateful-processing.md]: Updated to today.

File: src/_posts/2017-02-09-stateful-processing.md (hunk 5 above)
Comment (kennknowles) [inline src/_posts/2017-02-09-stateful-processing.md]: Done. No more prepend anywhere in the post.

File: src/_posts/2017-02-09-stateful-processing.md (hunk 3 above)
Comment (kennknowles) [inline src/_posts/2017-02-09-stateful-processing.md]: done

File: src/_posts/2017-02-09-stateful-processing.md (hunk 4 above)
Comment (kennknowles) [inline src/_posts/2017-02-09-stateful-processing.md]: done

File: src/_posts/2017-02-09-stateful-processing.md (hunk 6 above)
Comment (kennknowles) [inline src/_posts/2017-02-09-stateful-processing.md]: I spent ~5 minutes in the Javascript to hack something but didn't get it working. I added Python placeholders so it wasn't mysteriously absent, and the user will have the call to action to switch to Java.

**Comment by davorbonaci** [comment]
Beautiful post! Awesome work @kennknowles.

LGTM. Merging.

**Pull request closed without merging**